### PR TITLE
Added earlystopping mechanism

### PR DESCRIPTION
**Pull request by antoniovilarinholopes**
# Early stopping

A very requested feature for training regularisation to avoid overfitting. Issue #1382 is an example of requesting this simple feature. 

This pull request uses the previous codebase for inspiration.

## Detailed changes

- Added option to opts.py regarding the number of validation steps without improving (tolerance);
- Added earlystopping callable to trainer.py (only for gpu rank==0);
- Added earlystopping logic: after validation deciding whether or not to stop training; added `onmt/utils/earlystopping.py` with all the logic. Design was made thinking about future metrics, should be fairly easy to add ROUGE or BLEU in the future. 

## Adding metrics for earlystopping
1. Add new scorer that extracts the metric from stats (TODO, add other metrics to stats during validation);
2. Register new scorer in `SCORER_BUILDER`.

## Running with new options:
- `... --early_stopping 4 --early_stopping_criteria accuracy ppl` : 4 tolerance and use both accuracy and perplexity as earlystopping criteria.

### CODE REVIEW

**Comment by vince62s** [comment]
Thanks for this, looks great.
I have not tested it but since it won't break anything, might be good to merge and have users test it.
@francoishernandez @guillaumekln do yo want ot have a look ?

**Comment by guillaumekln** [comment]
What happens in the case of multi-GPU training? The master worker should somehow send a message to the other workers to terminate their training loop.

**Comment by vince62s** [comment]
Well you are correct, but:
If my understanding is correct, in the current codebase, we perform the validation on each spawn process (whether this is wanted or not is another story) which means that the valid task is the same on easch process with the "gathered" stats. So maybe the easiest way is to remove the test gpu_rank==0 in the callable definition.
Am I correct ?

**Comment by antoniovilarinholopes** [comment]
@guillaumekln you are correct, I forgot sending the other workers a stopping signal. 
@vince62s suggestion should solve it, I think.

**Comment by vince62s** [comment]
ok thanks, merging.

**Comment by chenqiuyuan** [comment]
Hi, I think this page should be integrated into the tutorial (I mean, the library http://opennmt.net/OpenNMT-py/Library.html). It takes me a lot of time to find this page before wasting a lot of time.
